source/server/listener_manager_impl.cc

```diff
@@ -558,6 +558,10 @@ void ListenerManagerImpl::drainListener(ListenerImplPtr&& listener) {
         // main thread to avoid locking. This makes sure that we don't destroy the listener
         // while filters might still be using its context (stats, etc.).
         server_.dispatcher().post([this, draining_it]() -> void {
+          // TODO(lambdai): Resolve race condition. The below refcount dec is too early. The active
+          // listener at worker thread could be destroyed before active connection.
+          // As a consequence, the listener at master thread and the connection at worker thread are
+          // destroyed concurrently.
```

Comment:
I don't follow the race condition. Clarify?

Reply:
Sorry for the unclear. I think I have an example in original PR.
Consider the below events in global sequence order
1. master thread: calling drainListener
2. work thread: *deferred delete* the active connection
3. work thread: post to master that the drain is done
4. master thread: erase the listener
5. worker thread: execute destroying connection when the shared listener config is destroyed at step 4 (could be worse such as access the connection because connection is not yet started to deleted).

The race condition is introduced because 3 occurs too early. My solution is to defer schedule the callback posting to master thread, by introducing DeferTaskUtil. So that 5 should always happen before 3.

Reply:
OK we can cover this in a later PR. Until then please update the comment to be more complete as you have written here.